### ISSUE TITLE
Hangfire dashboard is unable to 'requeue' job on other then default JobStorage

### ISSUE THREAD

**Issue by dzerzi** [HangfireIO/Hangfire#1831]
Hi,

We are using Hangfire Dashboard in a multitenant app where each tenant have it's own server. We want to expose one Dashboard and that working perfectly fine with UseHangfireDashboard & MapHangfireDashboard extensions but as soon as we want to use 'requeue' action on other tenant then that one used in 'AddHangfire' extension (btw. we shouldn't been forced to use that extension when we expose only Dashboard on app) we are trying to use JobStorage used in AddHangfire instead of JobStorage provided in UseHangfireDashboard.

To summarize:
- We see all jobs for choosed tenant on all dashboards:
- http://hf/tenantAlpha  -- will show only tenantAlpha jobs
- http://hf/tenantBeta -- will show only tenantBeta jobs

however when we trigger 'reque' on 'tenantAlpha' which is used in AddHangfire extension it requeue just fine, doing same on Beta do... exactly the same, including using same connectionString which is not welcome on multi tenant env ;-)

**Comment by dzerzi** [comment]
Ok, following suggestions from another 'bug' I've prepared this:

`        protected override void RegisterCustomImplementations(ContainerBuilder builder)
        {
            builder.Register(b => GetJobStorage( b.Resolve<ILifetimeScope>())).As<JobStorage>();

            builder.RegisterType<BackgroundJobStateChanger>().AsImplementedInterfaces().InstancePerLifetimeScope();
            builder.RegisterType<BackgroundJobClient>().AsImplementedInterfaces().InstancePerLifetimeScope();
        }

        private static SqlServerStorage GetJobStorage(ILifetimeScope scope)
        {
            var tenant = GetTenantFromContext(scope);
            var connectionString = GetConnectionString(tenant);
            return new SqlServerStorage(connectionString, HangFireConstants.SqlServerStorageOptions);
        }

        private static string GetTenantFromContext(ILifetimeScope scope)
        {
            var databases = GetTenants(); 
            var ctx = scope.Resolve<IHttpContextAccessor>();
            var path = ctx.HttpContext.Request.Headers[":path"].ToString();
            return path.Split("/").Single(databases.Contains);
        }`

We are using Autofac (so in fact that 3 classes is now scoped not singletons)

**Comment by dzerzi** [comment]
Ok, unfortunetly there is same issue with triggering recurring jobs. It remember first used tenant and register jobFactory per lifetime does not fix it - while register recurringJobManager even crashes. 

`            builder.RegisterType<BackgroundJobFactory>().AsImplementedInterfaces().InstancePerLifetimeScope();
            //builder.RegisterType<RecurringJobManager>().AsImplementedInterfaces().InstancePerLifetimeScope();`

**Comment by odinserj** [comment]
Thanks for reporting this, I've just added [`IBackgroundJobClientFactory`](https://github.com/HangfireIO/Hangfire/blob/master/src/Hangfire.AspNetCore/IBackgroundJobClientFactory.cs#L21) and [`IRecurringJobManagerFactory`](https://github.com/HangfireIO/Hangfire/blob/9605fc8fa29554ea178ddbe9f24bf529faff7b12/src/Hangfire.AspNetCore/IRecurringJobManagerFactory.cs#L21) interfaces you can register to use a client for the same storage as current dashboard page. They provide the following methods:

```csharp
IBackgroundJobClient GetClient(JobStorage storage);
IRecurringJobManager GetManager(JobStorage storage);
```

Changes will be released with Hangfire 1.7.20 today.

**Comment by odinserj** [comment]
Ah, everything will be even working automatically in version 1.7.20 without any changes from your side – I've added a default implementation for those interfaces that will be registered automatically.
